### PR TITLE
Cancellation refactor 4: Update tests for the new retry-cancellation api

### PR DESCRIPTION
**Pull request by e00E**
This is the final PR and should pass CI.

### Test Plan
New unit tests, CI.

### CODE REVIEW

File: services-core/src/solution_submission/retry.rs [hunk 1]
```diff
@@ -399,26 +414,61 @@ mod tests {
...
             .return_once(|_, _, _, _, _| {
                 async move {
-                    receiver.await.unwrap();
+                    submit_receiver.await.unwrap();
                     Ok(())
                 }
                 .boxed()
             });
-        sleep.expect_sleep().times(1).returning(|_| immediate!(()));
+        sleep.expect_sleep().return_once(move |_| {
```
Comment (fleupold) [inline services-core/src/solution_submission/retry.rs:422]: I'm not a big fan of using the sleep method here to kickstart the future resolution as it relies on pretty detailed knowledge of the intrinsics (e.g. the fact that sleep is called after the first solution submission).

Would it work to `cancel_sender.send(()).unwrap();` inside `expect_submit_solution` before `submit_receiver.await.unwrap()`? This way the test would say: we cancel as soon as we have issued the first tx and resolve the first tx as soon as we have started the cancellation.

File: services-core/src/solution_submission/retry.rs [hunk 2]
```diff
@@ -399,26 +414,61 @@ mod tests {
...
             .times(1)
-            .return_once(|_, _, _, _, _| {
-                sender.send(()).unwrap();
-                immediate!(Err(nonce_error()))
-            });
-        sleep
-            .expect_sleep()
-            .returning(|_| future::pending().boxed());
+            .return_once(|_, _, _, _, _| future::pending().boxed());
+        sleep.expect_sleep().return_once(move |_| {
```
Comment (fleupold) [inline services-core/src/solution_submission/retry.rs:464]: ditto: if possible I'd move the cancellation signalling into the `expect_submit_solution` mock.

**Review by fleupold — APPROVED** [review]
Good new tests!

File: services-core/src/solution_submission/retry.rs (hunk 1 above)
Comment (e00E) [reply services-core/src/solution_submission/retry.rs:422]: Good point, changed.

**Review by nlordell — APPROVED** [review]
Nice! The split up PRs made this much easier to review.

**Pull request merged**